cache.py

```diff
@@ -0,0 +1,92 @@
+import hashlib
+import inspect
+import os
+import pathlib
+from functools import wraps
+
+import appdirs
+import daiquiri
+from diskcache import Cache
+
+logger = daiquiri.getLogger("cache")
+
+
+def get_associated_files(func):
+    root_directory = os.path.dirname(inspect.getfile(func))
+
+    def should_include(func):
+        return (
+            os.path.commonpath([os.path.dirname(inspect.getfile(func)), root_directory])
+            == root_directory
+        ) and "site-packages" not in inspect.getfile(func)
+
+    def recursive_get_files(func, seen=None):
+        if seen is None:
+            seen = set()
+        if func in seen:
+            return set()
+        seen.add(func)
+        file_paths = set()
+        if should_include(func):
+            file_paths.add(inspect.getfile(func))
+
+        module = inspect.getmodule(func)
+        for _name, obj in inspect.getmembers(module):
+            if inspect.isclass(obj):
+                for _attr_name, attr_value in obj.__dict__.items():
+                    if isinstance(attr_value, property) and should_include(
+                        attr_value.fget
+                    ):
+                        file_paths |= recursive_get_files(attr_value.fget, seen)
+            if inspect.isfunction(obj) and should_include(obj):
+                file_paths |= recursive_get_files(obj, seen)
+
+        return file_paths
+
+    return recursive_get_files(func)
+
+
+def hash_files(file_paths):
+    m = hashlib.sha256()
+    for path in sorted(file_paths):  # Sort to ensure consistent hash
+        with open(path, "rb") as f:
+            m.update(f.read())
+    return m.hexdigest()
+
+
+def function_hash(func):
+    files = get_associated_files(func)
+    logger.info(f"Hashing {func.__name__} with files {files}")
+    return hash_files(get_associated_files(func))
+
+
+def get_cache_dir():
+    cache_dir = pathlib.Path(appdirs.user_cache_dir("tsqc", "tsqc"))
+    try:
+        os.makedirs(cache_dir)
+        logger.info(f"Set cache_dir to {cache_dir}")
+    except OSError:
```

Comment:
This could also happen, e.g. if there wasn't sufficient permissions, or if it already existed and was a file, right? So, the log error will be misleading.

Why not use the pathlib API directly?
```
cache_dir.mkdir(exist_ok=True, parents=True)
```
I think this takes care of all the required corner cases and raises a sensible error on weirdnesses.